File: src/main/java/pl/com/tegess/controller/user/readcontroller/UserReadController.java
package pl.com.tegess.controller.user.readcontroller;

import org.bson.types.ObjectId;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import org.springframework.web.bind.annotation.*;
import pl.com.tegess.controller.user.request.UserData;
import pl.com.tegess.domain.user.UserRepository;

import java.util.List;
import java.util.stream.Collectors;

@RestController
@Component
public class UserReadController {

    @Autowired
    UserRepository repository;

    @RequestMapping(value = "/api/users", method = RequestMethod.GET)
    public List<UserData> getUsersData() {

        return repository.findAll()
                .stream()
                .map(user -> new UserData(user.getId().toString(), user.getUsername(), user.getMail(), user.getPictureURL()))
                .collect(Collectors.toList());

    }

    @RequestMapping(value = "/api/applications/{appId}/users", method = RequestMethod.GET)
    public List<UserData> getUserDataForApplication(@PathVariable String appId) {
        return repository.findAll()
                .stream()
                .filter(user -> user.getAppId().equals(new ObjectId(appId)))
                .map(user -> new UserData(
                        user.getId().toString(),
                        user.getUsername(),
                        user.getMail(),
                        user.getPictureURL()))
                .collect(Collectors.toList());
    }

 }

File: src/main/scala/com/tegess/controller/user/UserReadController.scala
package com.tegess.controller.user

import com.tegess.controller._
import com.tegess.controller.user.UserReadController.UserData
import com.tegess.persistance.service.user.UserService
import org.bson.types.ObjectId
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.stereotype.Component
import org.springframework.web.bind.annotation.{PathVariable, RequestMapping, RequestMethod, RestController}

import scala.util.{Failure, Success, Try}


@RestController
@Component
class UserReadController {

  @Autowired var userService:UserService = _

  @RequestMapping(value=Array("/api/applications/{appId}/users"), method=Array(RequestMethod.GET))
  def getUsers(@PathVariable appId: String) = {
    val id = new ObjectId(appId)

    val users = userService.findAll(id)

    for {
      user <- users
    } yield UserData(user.username, user.username, user.mail, user.pictureUrl,
      user.gender, user.locale.map(_.toLanguageTag), user.privileges)

  }

  @PreAuthorize(value = "authentication.getName() == #appId.concat(#userId)")
  @RequestMapping(value=Array("/api/applications/{appId}/users/{userId}"), method=Array(RequestMethod.GET))
  def getUser(@PathVariable appId: String, @PathVariable userId: String) = {
    val userTry = for {
      applicationId <- Try(new ObjectId(appId))
      user <- userService.findOne(applicationId, userId).toTry
    } yield UserData(user.username, user.username, user.mail, user.pictureUrl,
      user.gender, user.locale.map(_.toLanguageTag), user.privileges)

    userTry match {
      case Success(u) => u
      case Failure(e) => throw new IllegalArgumentException("Something went wrong", e)
    }

  }

}
object UserReadController {
  case class UserData(id: String,
                      username: String,
                      mail: Option[String],
                      pictureURL: Option[String],
                      gender: Option[String],
                      locale: Option[String],
                      privileges: List[String])
}